fdbserver/GrvProxyServer.actor.cpp

```diff
@@ -302,6 +304,116 @@ ACTOR Future<Void> healthMetricsRequestServer(GrvProxyInterface grvProxy,
+					Tuple size = Tuple().append(sizeLimitInt);
+					tr->set(GlobalConfig::prefixedKey(fdbClientInfoTxnSizeLimit), size.pack());
+				}
+
+				wait(tr->commit());
+				break;
+			} catch (Error& e) {
+				// Multiple GRV proxies may attempt this migration at the same
+				// time, sometimes resulting in aborts due to conflicts.
+				TraceEvent(SevInfo, "GlobalConfigRetryableMigrationError").errorUnsuppressed(e).suppressFor(1.0);
+				wait(tr->onError(e));
+			}
+		}
+	} catch (Error& e) {
+		// Catch non-retryable errors (and do nothing).
+		TraceEvent(SevWarnAlways, "GlobalConfigMigrationError").error(e);
+	}
+	return Void();
+}
+
+// Periodically refresh local copy of global configuration.
+ACTOR Future<Void> globalConfigRefresh(GrvProxyData* grvProxyData, Version* cachedVersion, RangeResult* cachedData) {
+	state Reference<ReadYourWritesTransaction> tr = makeReference<ReadYourWritesTransaction>(grvProxyData->cx);
+	loop {
+		try {
+			tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+			state Optional<Value> globalConfigVersion = wait(tr->get(globalConfigVersionKey));
+			RangeResult tmpCachedData = wait(tr->getRange(globalConfigDataKeys, CLIENT_KNOBS->TOO_MANY));
+			*cachedData = tmpCachedData;
+			if (globalConfigVersion.present()) {
+				Version parsedVersion;
+				memcpy(&parsedVersion, globalConfigVersion.get().begin(), sizeof(Version));
+				*cachedVersion = bigEndian64(parsedVersion);
+			}
+			return Void();
+		} catch (Error& e) {
+			wait(tr->onError(e));
+		}
+	}
+}
+
+// Handle common GlobalConfig transactions on the server side, because not all
+// clients are allowed to read system keys. Eventually, this could become its
+// own role.
+ACTOR Future<Void> globalConfigRequestServer(GrvProxyData* grvProxyData, GrvProxyInterface grvProxy) {
+	state ActorCollection actors(false);
+	state Future<Void> refreshFuture; // so there is only one running attempt
+	state Version cachedVersion = 0;
+	state RangeResult cachedData;
+	state Future<Void> refreshTimeout = delay(SERVER_KNOBS->GLOBAL_CONFIG_REFRESH_INTERVAL);
+
+	// Run one-time migration to support upgrades.
+	wait(success(timeout(globalConfigMigrate(grvProxyData), SERVER_KNOBS->GLOBAL_CONFIG_MIGRATE_TIMEOUT)));
+
+	loop {
+		choose {
+			when(GlobalConfigRefreshRequest refresh = waitNext(grvProxy.refreshGlobalConfig.getFuture())) {
+				// Must have an up to date copy of global configuration in
+				// order to serve it to the client (up to date from the clients
+				// point of view. The client learns the version through a
+				// ClientDBInfo update).
+				if (refresh.lastKnown <= cachedVersion) {
+					refresh.reply.send(GlobalConfigRefreshReply{ cachedData.arena(), cachedData });
+				} else {
+					refresh.reply.sendError(timed_out());
+				}
+			}
+			when(wait(refreshTimeout)) {
+				refreshFuture = globalConfigRefresh(grvProxyData, &cachedVersion, &cachedData);
+				refreshTimeout = delay(SERVER_KNOBS->GLOBAL_CONFIG_REFRESH_INTERVAL);
```

Comment:
If for some reason the config refresh is taking longer than 1s, this is going to repeatedly cancel the read and start a new one, meaning that updates will stop until one is able to complete in under 1s.

Reply:
How about doing something like wait on either the refresh timeout (1 second) or the `refreshFuture` timing out (say with a ten second timeout). So if the refresh takes less than a second, we still wait a second, but if the refresh takes five seconds, we would wait five seconds. If the refresh takes more than ten seconds, we cancel it and start a new read.

Reply:
That sounds good. Another formulation of this could be:

```
refreshFuture = timeout(globalConfigRefresh(...), 10) && delay(SERVER_KNOBS->GLOBAL_CONFIG_REFRESH_INTERVAL);
```

Reply:
Thanks, that's much cleaner